src/server.rs

```diff
@@ -0,0 +1,221 @@
+    SelectCharacter,    // CC
+    ICMessage,          // MS
+    OOCMessage,         // CT
+    PlaySong,           // MC
+    WTCEButtons,        // RT
+    SetCasePreferences, // SETCASE
+    CaseAnnounce,       // CASEA
+    Penalties,          // HP
+    AddEvidence,        // PE
+    DeleteEvidence,     // DE
+    EditEvidence,       // EE
+    CallModButton,      // ZZ
+    KickWithGuard,      // opKICK
+    BanWithGuard,       // opBAN
+}
+
+impl FromStr for Command {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        use Command::*;
+        match s {
+            "HI" => Ok(Handshake),
+            "ID" => Ok(ClientVersion),
+            "CH" => Ok(KeepAlive),
+            "askchaa" => Ok(AskListLengths),
+            "askchar" => Ok(AskListCharacters),
+            "AN" => Ok(CharacterList),
+            "AE" => Ok(EvidenceList),
+            "AM" => Ok(MusicList),
+            "RC" => Ok(AO2CharacterList),
+            "RD" => Ok(AO2Ready),
+            "CC" => Ok(SelectCharacter),
+            "MS" => Ok(ICMessage),
+            "CT" => Ok(OOCMessage),
+            "MC" => Ok(PlaySong),
+            "RT" => Ok(WTCEButtons),
+            "SETCASE" => Ok(SetCasePreferences),
+            "CASEA" => Ok(CaseAnnounce),
+            "HP" => Ok(Penalties),
+            "PE" => Ok(AddEvidence),
+            "DE" => Ok(DeleteEvidence),
+            "EE" => Ok(EditEvidence),
+            "ZZ" => Ok(CallModButton),
+            "opKICK" => Ok(KickWithGuard),
+            "opBAN" => Ok(BanWithGuard),
+            _ => anyhow::bail!("Invalid command!"),
+        }
+    }
+}
+
+enum DecodeState {
+    Command,
+    Arguments,
+}
+
+use serde::export::Formatter;
+use std::{
+    io::{BufRead, Cursor, Seek, SeekFrom},
+    str::FromStr,
+};
+use tokio::{
+    io::{AsyncBufReadExt, Error, ErrorKind},
+    stream::StreamExt,
+};
+use tokio_util::codec::{Decoder, Encoder, FramedRead};
+
+#[derive(Debug)]
+pub struct AOMessage {
+    pub command: Command,
+    pub args: Vec<String>,
+}
+
+pub struct AOMessageCodec {
+    state: DecodeState,
+}
+
+impl AOMessageCodec {
+    pub fn new() -> Self {
+        Self { state: DecodeState::Command }
+    }
+}
+
+impl Decoder for AOMessageCodec {
+    type Item = AOMessage;
+    type Error = anyhow::Error;
+
+    fn decode(
+        &mut self,
+        mut src: &mut BytesMut,
+    ) -> Result<Option<Self::Item>, Self::Error> {
+        let mut reader = Cursor::new(src);
+        let mut cmd_buf = Vec::with_capacity(8);
+        let cmd_len = BufRead::read_until(&mut reader, b'#', &mut cmd_buf)?;
+
+        if cmd_len == 0 {
+            log::debug!("Invalid protocol?!");
+            anyhow::bail!("Invalid protocol!")
+        }
+
+        cmd_buf.truncate(cmd_len.saturating_sub(1));
+        let command: Command = unsafe {
+            // We are sure that the commands will *always* be valid utf8
+            String::from_utf8_unchecked(cmd_buf)
+        }
```

Comment:
You are **not** sure. The client can do whatever he wants and this is **UB**.
```suggestion
        let command: Command = unsafe {
            // We are sure that the commands will *always* be valid utf8
            String::from_utf8(cmd_buf).expect("Expected valid utf-8")
        }
```
(though, probably we want to just ignore invalid utf-8 instead of crashing)